node/src/reactor/main_reactor/keep_up.rs

```diff
@@ -464,9 +628,32 @@ impl MainReactor {
                         // even if we don't have a complete block (all parts and dependencies)
                         // we may have the parent's block header; if we do we also
                         // know its era which allows us to know if we have the validator
-                        // set for that era or not
+                        // set for that era or not;
+                        // note: there is a special case here where the parent might be an
+                        // immediate switch block - we check for that case by attempting to read
+                        // its parent and seeing whether it is also a switch block; if it is, we
+                        // pass the parent metadata on in the Sync instruction, so that we can read
+                        // the correct set of validators if the validators matrix doesn't have the
+                        // validators for the parent's era yet
+                        let maybe_parent_metadata = self
+                            .storage
+                            .read_block_header(parent_block_header.parent_hash())
+                            .map_err(|err| err.to_string())?
+                            .and_then(|parent_parent_header| {
```

Comment:
`grandparent`? ;-)

Reply:
I guess it might be more readable :wink: